StructuralMechanics: Process for comparing the eigenvalues postprocessing files

philbucher opened the pull request:
The test `Eigen_Q4_Thick_2x2_Plate_test` failes bcs I added the file postprocessing checking (introduced in #1079 ) with plain file comparison.
This does not work bcs of machine accuracies so I wrote a process that checks the values.
I checked this on two machines, now it should work fine 

Feast is needed to run this test in case you try (this test is skipped if feast is not available)

philbucher committed:
Initial commit for comparison process

philbucher committed:
final adjustments

philbucher committed:
typo

philbucher committed:
minor

philbucher commented:
Hi @msandre I think you also noticed that the test failed, could you please try if it works for you? Thx

loumalouomega reviewed (dismissed):
Ok for me

philbucher commented:
Vicente did you run the test?

loumalouomega commented:
![imagen](https://user-images.githubusercontent.com/19991680/34668198-90569f5c-f46c-11e7-94eb-189407059a47.png)

loumalouomega commented:
Maybe we should add .res to the gitignore to avoid having new files detected by git

msandre commented on applications/StructuralMechanicsApplication/python_scripts/check_postprocess_eigenvalues_file_process.py:
```diff
@@ -0,0 +1,123 @@
+                # abs bcs the sign does not matter for eigenvectors
+                # tolerance is small bcs values in file are also small
+                self.assertAlmostEqual(abs(float(tmp1[j])),
+                                       abs(float(tmp2[j])), 15)
```
It would be good if you could make the tolerance an input parameter instead of hard coding it.

msandre commented on applications/StructuralMechanicsApplication/python_scripts/check_postprocess_eigenvalues_file_process.py:
```diff
@@ -0,0 +1,123 @@
+        ref_file = open(ReferenceFileName,'r')
+        out_file = open(OutputFileName,'r')
+        
+        lines1 = ref_file.readlines()
```
Could you name these variables something like ref_lines and output_lines or check_lines?

philbucher commented:
thx @loumalouomega 
Yep adding .res to gitignore makes sense in my eyes.

I also wanted to add a process that deletes files, e.g. after a test

@loumalouomega what is this new code review tool? should we follow it or what is the rule on this?

loumalouomega commented:
We should, @pooyan-dadvand says

msandre commented on applications/StructuralMechanicsApplication/python_scripts/check_postprocess_eigenvalues_file_process.py:
```diff
@@ -0,0 +1,123 @@
+            if len(tmp1) != len(tmp2):
+                raise Exception("Different number of results!")
+
+            for j in range(len(tmp1)):
```
Code readability could be improved here by doing something like 

```
for ref_value, output_value in zip(list_of_ref_values, list_of_output_values):
    self.assertAlmostEqual(...)
```

msandre commented on applications/StructuralMechanicsApplication/python_scripts/check_postprocess_eigenvalues_file_process.py:
```diff
@@ -0,0 +1,123 @@
+        for i in range(num_lines_1):
+            if lines1[i].startswith("Result "):
+                results_start_index = i
+                results_available = True
```
I think `results_found` would be better here.

msandre commented on applications/StructuralMechanicsApplication/python_scripts/check_postprocess_eigenvalues_file_process.py:
```diff
@@ -0,0 +1,123 @@
+
+class CheckPostprocessEigenvaluesFileProcess(KratosMultiphysics.Process, KratosUnittest.TestCase):
+    """
+    This process compares the *.post.res file written by the PostprocessEigenvaluesProcess
```
This class assumes a certain results format. Can you give a ref to the format or to the class that creates the results file expected by this class?

pooyan-dadvand commented:
The codacy is an static analysis tool and I am testing its output on Kratos. At this moment is not forced to comply it to be merge-able but of course I would encourage you to take a look at its reports. 

Nevertheless some tuning of its rules are necessary to be practical in Kratos. Of course all your comments are welcomed and I will make an issue for that.

philbucher commented:
Hi @pooyan-dadvand thx for the explanation
I completely agree with using this, one improvement I am not sure abt is [this line](https://github.com/KratosMultiphysics/Kratos/pull/1247/files#diff-ed4eee0fbd7d8e122d2a8a4f643beaf6R12). The type checking is used everywhere, I didn't come up with this. In this case I think type checking is actually better, unless there will be derived classes of `Parameters`

Commit:
Merge branch 'master' into StructMech/compare-eigen-post-files

philbucher committed:
Merge branch 'master' into StructMech/compare-eigen-post-files

philbucher committed:
Merge branch 'master' into StructMech/compare-eigen-post-files

philbucher committed:
Removing not needed stuff

philbucher committed:
Added new test

philbucher committed:
Updated test to be self contained

philbucher committed:
Improved comparison process

philbucher committed:
Updated tests to use new format

philbucher commented:
Hi guys,
I finally had time to finish this, sorry it took so long.

I changed how I test the process, now the test is self-contained

I implemented the comparison of `.post.res` files in a generic way in the `CompareTwoFilesCheckProcess` such that it can also be used by others

@loumalouomega I improved this process a bit and slightly changed the interface. I updated the tests in the MeshingApplication, but I cannot run them because I don't have MMG installed. Could you please run them and see if they are still working? (no need to recompile I only changed python files)

philbucher committed:
removed parameter to check default value

loumalouomega reviewed (approved):
For me OK, it is clearer, sorry if I chose a bad name :P
